### PR TITLE
New plan is not immediately accessible on the planning tool

### LINKED ISSUE #1412: New plan is not immediately accessible on the planning tool
When you create a new plan it is not immediately accessible on the planning tool unless you refresh the page.

Need to fix this.  Perhaps look into optimistically making the plan available in the UI.

### PR DESCRIPTION
**Pull request by haseemVD**
This PR closes #1412

### CODE REVIEW

File: src/components/forms/PlanForm/index.tsx [hunk 1]
```diff
@@ -335,6 +346,12 @@ const PlanForm = (props: PlanFormProps) => {
         .create(payload)
         .then(() => {
           onSubmitSuccess(setSubmitting, setAreWeDoneHere, payload, addPlan);
+          const record = extractPlanRecordResponseFromPlanPayload(payload);
+          if (!record) {
+            return;
+          }
```
Comment (moshthepitt) [inline src/components/forms/PlanForm/index.tsx:352]: Would it make this code simpler to have the condition as `if (record)` and then avoid the `return`?

File: src/components/forms/PlanForm/index.tsx [hunk 2]
```diff
@@ -1261,7 +1278,7 @@ const PlanForm = (props: PlanFormProps) => {
   );
 };
 
-export const defaultProps: PlanFormProps = {
+export const defaultProps: Omit<PlanFormProps, 'actionCreator'> = {
```
Comment (moshthepitt) [inline src/components/forms/PlanForm/index.tsx:1281]: why is it necessary to change this type?

Comment (haseemVD) [reply src/components/forms/PlanForm/index.tsx]: updated.

File: src/components/forms/PlanForm/index.tsx [hunk 3]
```diff
@@ -173,6 +182,8 @@ export type LocationChildRenderProp = (
 
 /** interface for plan form props */
 export interface PlanFormProps {
+  actionCreator: ActionCreator<FetchPlanRecordsAction>;
+  plansArray: PlanRecord[];
```
Comment (moshthepitt) [inline src/components/forms/PlanForm/index.tsx:186]: How are you using this plans array?  I don't see why PlanForm needs to receive an array of plans?

**Review by moshthepitt — CHANGES_REQUESTED** [review]

File: src/components/forms/PlanForm/index.tsx (hunk 3 above)
Comment (haseemVD) [reply src/components/forms/PlanForm/index.tsx:186]: @moshthepitt, On submitting new plan, We need this array so we can push new plan in it and then update the store.

Comment (moshthepitt) [reply src/components/forms/PlanForm/index.tsx:186]: Okay cool.  I think we don't need the entire array.  The store would work fine if we added just one plan definition.  Either by using an array of one element, or by adding an action creator that accepts a single plan.

File: src/components/forms/PlanForm/index.tsx [hunk 4]
```diff
@@ -1,1323 +1,1338 @@
...
+          displayError(e, AN_ERROR_OCCURRED, false);
+        });
+    } else {
+      apiService
+        .create(payload)
+        .then(() => {
+          onSubmitSuccess(setSubmitting, setAreWeDoneHere, payload, addPlan);
+          const record = extractPlanRecordResponseFromPlanPayload(payload);
+          if (record) {
+            const extractedPlanRecords = [record].filter(plan => !!plan);
```
Comment (ciremusyoka) [inline src/components/forms/PlanForm/index.tsx:349]: Is this line necessary?

**Review by ciremusyoka — DISMISSED** [review]

File: src/components/forms/PlanForm/index.tsx [hunk 5]
```diff
@@ -1,1323 +1,1337 @@
-import FormikEffect from '@onaio/formik-effect';
```
Comment (moshthepitt) [inline src/components/forms/PlanForm/index.tsx:1]: What is changing in this file?

File: src/containers/pages/InterventionPlan/NewPlan/General/index.tsx [hunk 6]
```diff
@@ -1,173 +1,184 @@
-import React, { useState } from 'react';
```
Comment (moshthepitt) [inline src/containers/pages/InterventionPlan/NewPlan/General/index.tsx:1]: What is changing in this file?

File: src/containers/pages/InterventionPlan/NewPlan/General/tests/index.test.tsx [hunk 7]
```diff
@@ -1,465 +1,480 @@
-import { mount, shallow } from 'enzyme';
```
Comment (moshthepitt) [inline src/containers/pages/InterventionPlan/NewPlan/General/tests/index.test.tsx:1]: What is changing in this file?

File: src/containers/pages/InterventionPlan/NewPlan/General/tests/index.intervention.test.tsx [hunk 8]
```diff
@@ -1,67 +1,81 @@
-import { mount, shallow } from 'enzyme';
```
Comment (moshthepitt) [inline src/containers/pages/InterventionPlan/NewPlan/General/tests/index.intervention.test.tsx:1]: What is changing in this file?

File: src/store/ducks/plans.ts [hunk 9]
```diff
@@ -76,21 +77,15 @@ export interface Plan extends PlanRecord {
   jurisdiction_path: string[];
 }
 
-/** UseContext - interface for PlanPayload.useContext[] items */
-export interface UseContext {
```
Comment (moshthepitt) [inline src/store/ducks/plans.ts:80]: Why is this deleted?

File: src/containers/pages/InterventionPlan/UpdatePlan/index.tsx [hunk 10]
```diff
@@ -173,7 +178,11 @@ const UpdatePlan = (props: RouteComponentProps<RouteParams> & UpdatePlanProps) =
       <h3 className="mb-3 page-title">{pageTitle}</h3>
       <Row>
         <Col md={8}>
-          <PlanForm {...planFormProps} addPlan={fetchPlan} />
+          <PlanForm
+            actionCreator={fetchPlanRecordsCreator}
+            {...planFormProps}
+            addPlan={fetchPlan}
```
Comment (moshthepitt) [inline src/containers/pages/InterventionPlan/UpdatePlan/index.tsx:184]: Does it make sense to now add all of these to `planFormProps` to keep things tidy?

File: src/containers/pages/InterventionPlan/NewPlan/General/tests/index.test.tsx [hunk 11]
```diff
@@ -217,20 +232,20 @@ describe('containers/pages/NewPlan', () => {
         .props() as any).formik.values
     );
 
-    expect(mockedStore.dispatch).toHaveBeenLastCalledWith(addPlanDefinition(payload));
+    expect(mockedStore.dispatch).toHaveBeenCalledWith(addPlanDefinition(payload));
```
Comment (moshthepitt) [inline src/containers/pages/InterventionPlan/NewPlan/General/tests/index.test.tsx:235]: Why does this change?  I'm trying to understand how this ensures that we are testing for the feature that you added.

File: src/containers/pages/InterventionPlan/NewPlan/General/tests/index.test.tsx [hunk 12]
```diff
@@ -322,21 +337,21 @@ describe('containers/pages/NewPlan', () => {
         .first()
         .props() as any).formik.values
     );
-
-    expect(mockedStore.dispatch).toHaveBeenLastCalledWith(addPlanDefinition(payload));
+    const payloadOfInterest = extractPlanRecordResponseFromPlanPayload(payload);
+    expect(mockedStore.dispatch).toHaveBeenLastCalledWith(fetchPlanRecords([payloadOfInterest]));
```
Comment (moshthepitt) [inline src/containers/pages/InterventionPlan/NewPlan/General/tests/index.test.tsx:341]: Could you explain this change?  Trying to understand the significance of swapping `addPlanDefinition(payload)` with `fetchPlanRecords([payloadOfInterest])`?

File: src/containers/pages/InterventionPlan/NewPlan/General/tests/index.test.tsx [hunk 13]
```diff
@@ -322,21 +337,21 @@ describe('containers/pages/NewPlan', () => {
...
   });
 
   it('New plan in planning tool is NOT added to store if API status is NOT 200', async () => {
     fetch.mockReject(() => Promise.reject('API is down'));
 
-    const initialState = {};
+    const initialState = existingState;
```
Comment (moshthepitt) [inline src/containers/pages/InterventionPlan/NewPlan/General/tests/index.test.tsx:347]: Why is this necessary?

File: src/containers/pages/InterventionPlan/NewPlan/General/tests/index.test.tsx [hunk 14]
```diff
@@ -443,9 +458,9 @@ describe('containers/pages/NewPlan', () => {
     });
     document.body.appendChild(div);
     const wrapper = mount(
-      <Provider store={mockStore({})}>
+      <Provider store={mockStore(existingState)}>
         <Router history={history}>
-          <ConnectedBaseNewPlan />
+          <Route component={ConnectedBaseNewPlan} />
```
Comment (moshthepitt) [inline src/containers/pages/InterventionPlan/NewPlan/General/tests/index.test.tsx:463]: Could you explain this change please?

File: src/containers/pages/InterventionPlan/NewPlan/General/tests/index.test.tsx [hunk 15]
```diff
@@ -217,20 +232,20 @@ describe('containers/pages/NewPlan', () => {
...
 
-    expect(mockedStore.dispatch).toHaveBeenLastCalledWith(addPlanDefinition(payload));
+    expect(mockedStore.dispatch).toHaveBeenCalledWith(addPlanDefinition(payload));
   });
 
   it('New plan is NOT added to store if API status is NOT 200', async () => {
     fetch.mockReject(() => Promise.reject('API is down'));
 
-    const initialState = {};
+    const initialState = existingState;
```
[the remark this answers is not in the document]
Comment (haseemVD) [reply src/containers/pages/InterventionPlan/NewPlan/General/tests/index.test.tsx:241]: we need to mock the data in the store because we dont have any data in start and this throws `undefined` when retrieving plans record from `store`

File: src/containers/pages/InterventionPlan/NewPlan/General/tests/index.test.tsx (hunk 13 above)
Comment (haseemVD) [reply src/containers/pages/InterventionPlan/NewPlan/General/tests/index.test.tsx:347]: we need to mock the data in the store because we dont have any data in start and this throws `undefined` when retrieving plans record from `store`

File: src/containers/pages/InterventionPlan/NewPlan/General/tests/index.test.tsx (hunk 14 above)
Comment (haseemVD) [reply src/containers/pages/InterventionPlan/NewPlan/General/tests/index.test.tsx:463]: component expect to be used with route props and that is only possible if we pass the components itself to the route.

File: src/store/ducks/plans.ts (hunk 9 above)
Comment (haseemVD) [reply src/store/ducks/plans.ts:80]: the reason of removing this interface was to remove the duplication of code we already have this interface defined [here](https://github.com/onaio/reveal-frontend/blob/master/src/configs/settings.ts#L1670) with better type check

File: src/containers/pages/InterventionPlan/NewPlan/General/tests/index.test.tsx (hunk 11 above)
Comment (haseemVD) [reply src/containers/pages/InterventionPlan/NewPlan/General/tests/index.test.tsx:235]: The order of dispatch changes. as the mocked store initial dispatches in initial state into the store itself thats why instead of testing for the last dispatch we can arbitrary check for weather it is called or not.

File: src/containers/pages/InterventionPlan/NewPlan/General/tests/index.test.tsx (hunk 14 above)
Comment (haseemVD) [reply src/containers/pages/InterventionPlan/NewPlan/General/tests/index.test.tsx:463]: component expect to be used with route props and that is only possible if we pass the components itself to the route

File: src/store/ducks/plans.ts [hunk 16]
```diff
@@ -76,21 +77,15 @@ export interface Plan extends PlanRecord {
...
-  code: string;
-  valueCodableConcept: string;
-}
-
 /** PlanPayload - interface for the payload used when creating/updating a plan via OpenSRP plans Endpoint */
 export interface PlanPayload {
-  action: any[];
+  action: PlanAction[];
```
Comment (moshthepitt) [inline src/store/ducks/plans.ts:82]: nice

File: src/components/forms/PlanForm/index.tsx [hunk 17]
```diff
@@ -335,6 +344,10 @@ const PlanForm = (props: PlanFormProps) => {
         .create(payload)
         .then(() => {
           onSubmitSuccess(setSubmitting, setAreWeDoneHere, payload, addPlan);
+          const record = extractPlanRecordResponseFromPlanPayload(payload);
```
Comment (moshthepitt) [inline src/components/forms/PlanForm/index.tsx:347]: Let's add a comment here explaining what we are doing.

**Review by moshthepitt — APPROVED** [review]

**Review by ciremusyoka — APPROVED** [review]
